crates/core/component/auction/src/component/dutch_auction.rs

```diff
@@ -51,11 +64,183 @@ pub(crate) trait DutchAuctionManager: StateWrite {
+
+        let DutchAuctionState {
+            sequence: _,
+            current_position,
+            next_trigger: _,
+            input_reserves: _,
+            output_reserves: _,
+        } = old_dutch_auction.state;
+
+        let auction_input_id = input.asset_id;
+        let auction_output_id = output_id;
+
+        let auction_trigger = TriggerData {
+            start_height,
+            end_height,
+            step_count,
+        };
+
+        // Recover the LP's balances, if it exists.
+        let lp_reserves = if let Some(auction_lp_id) = current_position {
+            self.close_position_by_id(&auction_lp_id)
+                .await
+                .map_err(|e| {
+                    tracing::error!(
+                        ?e,
+                        ?auction_lp_id,
+                        ?auction_id,
+                        "failed to close dutch auction LP"
+                    )
+                })
+                .expect("position should exist and be opened or closed");
+            self.withdraw_position(auction_lp_id, 0u64)
+                .await
+                .map_err(|e| {
+                    tracing::error!(
+                        ?e,
+                        ?auction_lp_id,
+                        ?auction_id,
+                        "failed to close dutch auction LP"
+                    )
+                })
+                .expect("no state incoherence")
+        } else {
+            Balance::zero()
+        };
+
+        // We remove the execution trigger that we are currently processing:
+        self.unset_trigger_for_dutch_id(auction_id, trigger_height);
+
+        // Prepare a new auction, based on the previous one.
+        let mut new_dutch_auction = DutchAuction {
+            description: old_dutch_auction.description,
+            state: old_dutch_auction.state,
+        };
+
+        let balance_input_asset = lp_reserves
+            .provided()
+            .filter(|v| v.asset_id == input.asset_id)
+            .map(|v| v.amount)
+            .sum::<Amount>();
+        let balance_output_asset = lp_reserves
+            .provided()
+            .filter(|v| v.asset_id == output_id)
+            .map(|v| v.amount)
+            .sum::<Amount>();
+
+        new_dutch_auction.state.current_position = None;
+        // We overwrite the input amount
+        new_dutch_auction.state.input_reserves = balance_input_asset;
+        new_dutch_auction.state.output_reserves += balance_output_asset;
+        new_dutch_auction.state.next_trigger = 0;
+
+        // Compute the current step index, between 0 and `step_count`.
+        let step_index = auction_trigger
+            .compute_step_index(trigger_height)
+            .expect("trigger data is validated");
+
+        if step_index >= step_count {
+            // If the termination condition has been reached, we set the auction
+            // sequence to 1 (Closed).
+            new_dutch_auction.state.sequence = 1;
```

Comment:
Not sure this termination condition is correct, I think it won't ever hit the reserve price, should it be `step_index > step_count`?

Reply:
That's right it should be strictly greater

Reply:
Ended up leaving it as is and changed the price interpolation logic so that we reach the minimum at `step_count - 1` (x-ref:  https://github.com/penumbra-zone/penumbra/pull/4254#discussion_r1575591551). I think this makes more sense to have the auction terminate at the specified `end_height`.